fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/protocol/TimeAverageMetricTransformer.java

```diff
@@ -0,0 +1,215 @@
+// Copyright 2020 Oath Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.data.metric.protocol;
+
+import static com.yahoo.bard.webservice.data.time.DefaultTimeGrain.DAY;
+import static com.yahoo.bard.webservice.data.time.DefaultTimeGrain.MONTH;
+import static com.yahoo.bard.webservice.data.time.DefaultTimeGrain.WEEK;
+
+import com.yahoo.bard.webservice.data.config.metric.makers.AggregationAverageMaker;
+import com.yahoo.bard.webservice.data.metric.LogicalMetric;
+import com.yahoo.bard.webservice.data.metric.LogicalMetricInfo;
+import com.yahoo.bard.webservice.data.metric.MetricDictionary;
+import com.yahoo.bard.webservice.data.time.ZonelessTimeGrain;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * An interface for transforming metrics into other metrics.
+ */
+public class TimeAverageMetricTransformer implements MetricTransformer {
+
+    private static final MetricDictionary EMPTY_METRIC_DICTIONARY = new MetricDictionary();
+
+    public static final String CONTRACT_NAME = "reagg";
+    private static final String CORE_PARAMETER = "reagg";
```

Comment:
This code is smelly to me. I don't have a full suggested change, and this code might be fine, but I think it should be discussed.

Protocol contracts are unique over an entire code base. You can't have multiple different Protocols with the same contract. These two fields are strongly coupling this transformer to a specific Protocol instance; the instance that takes the reagg contract with the reagg core parameter and uses this MetricTransformer as its transformer instance. However, that instance is never materialized anywhere in the code base. To use this, the client needs to materialize the specific Protocol that we are essentially defining for them.

However, the obvious issue is that the materialized Protocol might want to use a "Multiplexing" transformer to support other reagg operations in addition to this one. It is uncomfortable to just create the Protocol that this Transformer requires because it might need to be tweaked or expanded. Multiplexing on it's own raises an issue; if this contract and core parameter are shared with another Protocol, how do we determine which transformer should handle a signal? I think we need to at least define the MultiplexingTransformer, and make it enforce uniqueness of core parameter names.

That also raises another issue: It is possible to handle multiple different Protocols with the same contract in the same system but different transformers. Since you can't compare Protocols or MetricTransformers (no equals method override) it is difficult at runtime to tell if Protocols are different. They might have the same contract and core parameter, but with no way to compare transformers they could be totally different.

Depending on what we decide, it might also be worth defining an enum or collection of Protocol implementations for default Protocols.

Reply:
Addressed by my comment on MetricTransformer.java